third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,248 @@
+  // %2 = convert %1 (tmp -> blocked)
+  //
+  // Let's consider tmp as blocked layout:
+  // sizePerThread = [1, 4], threadsPerWarp = [32, 2], warpsPerCTA = [1, 4]
+  // Tmp layout scratch buffer has shape: [1*32*1, 4*2*4] = [32, 32]
+  //
+  // With intermediate layout we have two scratch buffers:
+  //
+  // %1 = convert %0 (mfma -> tmp): [max(32, 32), max(128, 32)] = [32, 128]
+  // %2 = convert %1 (tmp -> blocked): [max(32, 128), max(32, 32)] = [128, 32]
+  //
+  // Both of these buffers are 4x times smaller than original one and their live
+  // times do not intersect, therefore this transformation lowers LDS
+  // consumption.
+  void tryFitCvtIntoLDS(triton::gpu::ConvertLayoutOp cvtOp, int targetLDSSize) {
+    OpBuilder builder(cvtOp);
+
+    auto srcType = cvtOp.getSrc().getType();
+    auto dstType = cvtOp.getType();
+
+    auto srcEnc = srcType.getEncoding();
+    auto dstEnc = dstType.getEncoding();
+
+    auto ctx = srcEnc.getContext();
+    auto rank = srcType.getShape().size();
+    unsigned numWarps = triton::gpu::getNumWarpsPerCTA(srcEnc);
+    auto warpSize = triton::gpu::getWarpSize(srcEnc);
+
+    // Find all possible shapes of WarpsPerCTA by finding all possible
+    // factorizations of numWarps. Pick shape for which both conversions in
+    // decomposition use LDS less than LDSSize and for which sum of LDS usage
+    // is minimal. If no such shape exists, do not decompose.
+    auto factorizedNumWarps =
+        mlir::triton::AMD::factorizePowerOf2(numWarps, rank);
+    // Create a list of temporary layouts
+    SmallVector<Attribute> tmpLayouts;
+    for (int i = 0; i < factorizedNumWarps.size(); i++) {
+      auto warpsPerCTA = factorizedNumWarps[i];
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(srcEnc, warpsPerCTA));
+      tmpLayouts.push_back(
+          mlir::triton::AMD::createTmpLayout(dstEnc, warpsPerCTA));
+      SmallVector<unsigned> elemsPerThread(rank, 1);
+      SmallVector<unsigned> threadsPerWarp(rank, 1);
+      threadsPerWarp[rank - 1] = warpSize / 8;
+      threadsPerWarp[rank - 2] = warpSize / threadsPerWarp[rank - 1];
+      auto order = triton::gpu::getOrder(srcEnc);
+      auto layoutCTA = triton::gpu::getCTALayout(srcEnc);
+      auto fallbackLayout = triton::gpu::BlockedEncodingAttr::get(
+          ctx, elemsPerThread, threadsPerWarp, warpsPerCTA, order, layoutCTA);
+      tmpLayouts.push_back(fallbackLayout);
+    }
+
+    unsigned minLDSUsage = 2 * LDSSize;
+    int minIdx = -1;
+    for (int i = 0; i < tmpLayouts.size(); i++) {
+      auto resources = mlir::triton::AMD::estimateResourcesForReplacement(
+          builder, cvtOp, tmpLayouts[i]);
+      // TODO analyze performance along with LDS consumption
+      if (resources.LDS < minLDSUsage) {
+        minLDSUsage = resources.LDS;
+        minIdx = i;
+      }
+    }
+
+    if (minIdx == -1 || minLDSUsage > targetLDSSize) {
+      return;
+    }
+
+    triton::gpu::ConvertLayoutOp tmpCvt;
```

Comment:
are we using this tmpCvt?

Reply:
Nope, will rewrite this part as done in `DecomposeUnsupportedConversions` pass.